### PR TITLE
WIP: Simple lines 

### PR DESCRIPTION
**Pull request by TimHal**
## Summary
WIP: Implementation of a simple line-drawing tool. 

## Known Issues / Things To Do

The CanvasObject is created, but not drawn properly. It will show up on the right side panel and not produce crashes or exceptions. The new PolyLine class inherits Goo.PolyLine (https://valadoc.org/goocanvas-2.0/Goo.CanvasPolyline.html).

**Problems:**

- Unclear how to use  CanvasPolyline.line  instead of CanvasPolyline . The former does seem to have the desired signature (x1,y1,x2,y2) but apparently is not a constructor. 

 - I am unable to compile the project with the debug flag in order to use Nemiver or gdb on it (which would help a lot). The ninja build chain does have --debug in it's ARGS, but so far I have not found a proper way to debug (even print-debugging does not work too well). If anybody has some pointers, that would be greatly appreciated. I have done the `/usr/bin/com.github.akiraux.akiraDevel --gtk-debug=all --gtk-no-debug=layout,interactive,resize --gdk-debug=all &> akira.log
` but these messages are only gtk related from what I have seen. 

- Edit: Linting does currently fail, I will fix all linting problems as soon as the functionality is there :)

## This PR implements the following **features**:
 
Drawing a line between two points. Both points should be movable by dnd. (WIP)
I hope to implement multiple lines/connected polygons after getting the first steps right on this one.

### CODE REVIEW

**Comment by Alecaddd** [comment]
Since this is a WIP, as you correctly wrote in the title, you should have created this as a Draft PR so we know we shouldn't review it: https://github.blog/2019-02-14-introducing-draft-pull-requests/

> I am unable to compile the project with the debug flag

Meson only flags code errors during compiling.
After you run `ninja`, you can launch Akira with the command `G_MESSAGES_DEBUG=Akira ./src/com.github.akiraux.akira` to see the full debug.
Throughout the application we use `debug()`, `warning()`, and `critical()` to spit out messages in the debug.

You also don't need to install Akira every time if the changes don't involve gsettings or different libraries (that's the main purpose of having a devel installation, but other than that we pretty much use the regular version).

> Linting does currently fail, I will fix all linting problems as soon as the functionality is there :)

We use vala-lint for local linting. You can install it and run it by following these instructions: https://github.com/vala-lang/vala-lint

File: src/Lib/Models/CanvasPolyLine.vala [hunk 1]
```diff
@@ -0,0 +1,142 @@
+/*
+* Copyright (c) 2019-2020 Alecaddd (https://alecaddd.com)
```
Comment (Alecaddd) [inline src/Lib/Models/CanvasPolyLine.vala:2]: This file was created in 2020, so no need the 2019

File: src/Lib/Models/CanvasPolyLine.vala [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+* Akira is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+
+* You should have received a copy of the GNU General Public License
+* along with Akira. If not, see <https://www.gnu.org/licenses/>.
+*
+* Authored by: Alessandro "alecaddd" Castellani <castellani.ale@gmail.com>
```
Comment (Alecaddd) [inline src/Lib/Models/CanvasPolyLine.vala:19]: You created this file, so you should list your name here instead of mine

**Comment by TimHal** [comment]
> Since this is a WIP, as you correctly wrote in the title, you should have created this as a Draft PR so we know we shouldn't review it: https://github.blog/2019-02-14-introducing-draft-pull-requests/

Alright, will do that next time. Could you change it to a draft PR or shall I re-open this?

> After you run `ninja`, you can launch Akira with the command `G_MESSAGES_DEBUG=Akira ./src/com.github.akiraux.akira` to see the full debug.

Thanks a lot, exactly what I was looking for!

**Comment by Alecaddd** [comment]
> Alright, will do that next time. Could you change it to a draft PR or shall I re-open this?

No problem, you can keep this open for simplicity, it was just a heads up for next time :smile: 
Unfortunately GitHub doesn't currently offer the option to convert an opened PR into a draft PR.

File: src/Lib/Models/CanvasPolyLine.vala [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        item_type = Models.CanvasItemType.POLYLINE;
+        id = Models.CanvasItem.create_item_id (this);
+        Models.CanvasItem.init_item (this);
+        if (artboard != null) {
+            connect_to_artboard ();
+        }
+
+        points = new Goo.CanvasPoints (2);
+        points.set_point  (0, 10, 10);
+        points.set_point  (1, 100, 100);
```
Comment (DevAlien) [inline src/Lib/Models/CanvasPolyLine.vala:112]: It took me a bit to figure this out. Basically everytime you want to change the line you have to pass it new CanvasPoints, you cannot set them afterwards, they want be displayed/updated.

```
var p = new Goo.CanvasPoints (2);
p.set_point  (0, 10, 10);
p.set_point  (1, 100, 100);
points = p;
```

File: src/Layouts/HeaderBar.vala [hunk 4]
```diff
@@ -259,6 +259,12 @@ public class Akira.Layouts.HeaderBar : Gtk.HeaderBar {
         var sub_separator = new Gtk.Separator (Gtk.Orientation.HORIZONTAL);
         sub_separator.margin_top = sub_separator.margin_bottom = 3;
 
+        var polyline = create_model_button (
+            _("Line"),
+            "shape-line-symbolic",
+            Akira.Services.ActionManager.ACTION_PREFIX +
+            Akira.Services.ActionManager.ACTION_RECT_TOOL);
```
Comment (albfan) [inline src/Layouts/HeaderBar.vala:266]: ACTION_LINE_TOOL

File: src/Lib/Models/CanvasPolyLine.vala (hunk 3 above)
Comment (albfan) [reply src/Lib/Models/CanvasPolyLine.vala:112]: you probably want here _x1,_y1 and _x2, _y2

File: src/Lib/Models/CanvasPolyLine.vala [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+        if (has_border) {
+            border_color = _border_color;
+            border_size = _border_size;
+        }
+
+        reset_colors ();
+    }
+
+    public void update_border () {
```
Comment (albfan) [inline src/Lib/Models/CanvasPolyLine.vala:140]: copy pasta from Rect?

**Review by albfan — CHANGES_REQUESTED** [review]
Take a look at https://github.com/akiraux/interactive-goocanvas/commit/7e74489de40e5f9e55a77d253afc65253f88ec09

Is a pet project where you can play easily with CanvasItems

**Review by Abdallah-Moh — APPROVED** [review]

**Comment by Alecaddd** [comment]
Closing this as considered abandoned.

**Pull request closed without merging**